cloudinit/net/dhcp.py

```diff
@@ -532,22 +530,132 @@ def parse_dhcp_server_from_lease_file(lease_file) -> Optional[str]:
+                ]
+            )
+            return self.parse_dhcp_lease_file(interface)
+        except subp.ProcessExecutionError as error:
+            LOG.debug(
+                "dhclient exited with code: %s stderr: %r stdout: %r",
+                error.exit_code,
+                error.stderr,
+                error.stdout,
+            )
+            raise NoDHCPLeaseError from error
+
+    @staticmethod
+    def parse_dhcpcd_lease(lease_dump: str, interface: str) -> List[dict]:
+        """parse the output of dhcpcd --dump
+
+        map names to the datastructure we create from dhclient
+
+        example dhcpcd output:
+
+        broadcast_address='192.168.15.255'
+        dhcp_lease_time='3600'
+        dhcp_message_type='5'
+        dhcp_server_identifier='192.168.0.1'
+        domain_name='us-east-2.compute.internal'
+        domain_name_servers='192.168.0.2'
+        host_name='ip-192-168-0-212'
+        interface_mtu='9001'
+        ip_address='192.168.0.212'
+        network_number='192.168.0.0'
+        routers='192.168.0.1'
+        subnet_cidr='20'
+        subnet_mask='255.255.240.0'
+        """
+
+        # create a dict from dhcpcd dump output
+        lease = dict(
+            [a.split("=") for a in lease_dump.replace("'", "").split()]
+        )
+
+        # this name is just different
+        lease["fixed-address"] = lease.pop("ip_address")
+
+        # this is expected by our code, can probably change that at some point
+        lease["interface"] = interface
+
+        # transform underscores to hyphens and return
+        return [{key.replace("_", "-"): value for key, value in lease.items()}]
+
+    @classmethod
+    def parse_dhcp_lease_file(cls, interface) -> List[Dict[str, Any]]:
+        """Return a list of leases
+
+        Return a list of dicts of dhcp options. Each dict contains key value
+        pairs a specific lease in order from oldest to newest.
+
+        @raises: InvalidDHCPLeaseFileError on empty or unparseable leasefile
+            content.
+        """
+        try:
+            return cls.parse_dhcpcd_lease(
+                subp.subp(
+                    [
+                        "dhcpcd",
+                        "--dumplease",
+                        interface,
+                    ],
+                    rcs=[0, 1],
+                ).stdout,
+                interface,
+            )
+
+        except subp.ProcessExecutionError as error:
+            LOG.debug(
+                "dhcpcd exited with code: %s stderr: %r stdout: %r",
+                error.exit_code,
+                error.stderr,
+                error.stdout,
+            )
+            raise NoDHCPLeaseError from error
 
 
 class Udhcpc(DhcpClient):
     client_name = "udhcpc"
 
-    def __init__(self):
-        self.udhcpc_path = subp.which("udhcpc")
-        if not self.udhcpc_path:
-            LOG.debug("Skip udhcpc configuration: No udhcpc command found.")
-            raise NoDHCPLeaseMissingUdhcpcError()
```

Comment:
We can also drop the exception handling of MissingUdhcpcError  in `select_dhcp_client and drop the definition of this exception now